tests/integration/test_cluster_discovery/test.py

```diff
@@ -0,0 +1,64 @@
+import pytest
+import time
+
+from helpers.cluster import ClickHouseCluster
+
+cluster = ClickHouseCluster(__file__)
+
+nodes = [
+    cluster.add_instance(
+        f'node{i}',
+        main_configs=['config/config.xml'],
+        stay_alive=True,
+        with_zookeeper=True
+    ) for i in range(5)
+]
+
+
+@pytest.fixture(scope="module")
+def start_cluster():
+    try:
+        cluster.start()
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def check_nodes_count_in_cluster(nodes, expected, cluster_name, *, retries=5):
+    """
+    Check nodes count in system.clusters for specified cluster
+    """
+
+    for retry in range(1, retries + 1):
+        nodes_cnt = [
+            int(node.query(f"SELECT count() FROM system.clusters WHERE cluster = '{cluster_name}'"))
+            for node in nodes
+        ]
+        if all(actual == expected for actual in nodes_cnt):
+            break
+
+        if retry != retries:
+            time.sleep(2 ** retry)
+    else:
+        raise Exception(f'Wrong nodes count in cluster: {nodes_cnt}, expected: {expected} (after {retries} retries)')
+
+
+def test_cluster_discovery_startup_and_stop(start_cluster):
+    """
+    Start cluster, check nodes count in system.clusters,
+    then stop/start some nodes and check that it (dis)appeared in cluster.
+    """
+
+    check_nodes_count_in_cluster([nodes[0], nodes[2]], len(nodes), 'test_auto_cluster')
+
+    nodes[1].stop_clickhouse(kill=True)
+    check_nodes_count_in_cluster([nodes[0], nodes[2]], len(nodes) - 1, 'test_auto_cluster')
+
+    nodes[3].stop_clickhouse()
+    check_nodes_count_in_cluster([nodes[0], nodes[2]], len(nodes) - 2, 'test_auto_cluster')
```

Comment:
Do you want to actually remove nodes from ZK on stop? Think about `ON CLUSTER` alters. These should be still queued to nodes even if a node is currently stopped/crashed.

Reply:
It depends on use-case, e.g. if we want to use cluster over shared storage we can perform hot addition/removing nodes from cluster and should care only about whole cluster state, but not about the state of individual nodes